src/protocol/libp2p/kademlia/query/get_providers.rs

```diff
@@ -0,0 +1,267 @@
+        query::{QueryAction, QueryId},
+        record::Key as RecordKey,
+        types::{ConnectionType, Distance, KademliaPeer, Key},
+    },
+    types::multiaddr::Multiaddr,
+    PeerId,
+};
+
+use std::collections::{BTreeMap, HashMap, HashSet, VecDeque};
+
+/// Logging target for the file.
+const LOG_TARGET: &str = "litep2p::ipfs::kademlia::query::get_providers";
+
+/// The configuration needed to instantiate a new [`GetProvidersContext`].
+#[derive(Debug)]
+pub struct GetProvidersConfig {
+    /// Local peer ID.
+    pub local_peer_id: PeerId,
+
+    /// Parallelism factor.
+    pub parallelism_factor: usize,
+
+    /// Query ID.
+    pub query: QueryId,
+
+    /// Target key.
+    pub target: Key<RecordKey>,
+}
+
+#[derive(Debug)]
+pub struct GetProvidersContext {
+    /// Query immutable config.
+    pub config: GetProvidersConfig,
+
+    /// Cached Kademlia message to send.
+    kad_message: Bytes,
+
+    /// Peers from whom the `QueryEngine` is waiting to hear a response.
+    pub pending: HashMap<PeerId, KademliaPeer>,
+
+    /// Queried candidates.
+    ///
+    /// These are the peers for whom the query has already been sent
+    /// and who have either returned their closest peers or failed to answer.
+    pub queried: HashSet<PeerId>,
+
+    /// Candidates.
+    pub candidates: BTreeMap<Distance, KademliaPeer>,
+
+    /// Found providers.
+    pub found_providers: Vec<KademliaPeer>,
+}
+
+impl GetProvidersContext {
+    /// Create new [`GetProvidersContext`].
+    pub fn new(config: GetProvidersConfig, in_peers: VecDeque<KademliaPeer>) -> Self {
+        let mut candidates = BTreeMap::new();
+
+        for candidate in &in_peers {
+            let distance = config.target.distance(&candidate.key);
+            candidates.insert(distance, candidate.clone());
+        }
+
+        let kad_message =
+            KademliaMessage::get_providers_request(config.target.clone().into_preimage());
+
+        Self {
+            config,
+            kad_message,
+            candidates,
+            pending: HashMap::new(),
+            queried: HashSet::new(),
+            found_providers: Vec::new(),
+        }
+    }
+
+    /// Get the found providers.
+    pub fn found_providers(self) -> Vec<KademliaPeer> {
+        // Merge addresses of different provider records of the same peer.
+        let mut providers = HashMap::<PeerId, HashSet<Multiaddr>>::new();
+        self.found_providers.into_iter().for_each(|provider| {
+            providers
+                .entry(provider.peer)
+                .or_default()
+                .extend(provider.addresses.into_iter())
+        });
+
+        // Convert into `Vec<KademliaPeer>`
+        let mut providers = providers
+            .into_iter()
+            .map(|(peer, addresses)| KademliaPeer {
+                key: Key::from(peer.clone()),
+                peer,
+                addresses: addresses.into_iter().collect(),
+                connection: ConnectionType::CanConnect,
```

Comment:
nit: Could we use `NotConnected` instead? I do remember I was planning to remove these at some point but got side-tracked :D

Reply:
Sure! Addressed